File: F#/chap07/Domain.fs
namespace Domain

open System

// 入力データ
type UnvalidatedOrder =
    { OrderId: string
      CustomerInfo: UnvalidatedCustomer
      ShippingAddress: UnvalidatedAddress }

and UnvalidatedCustomer = { Name: string; Email: string }
and UnvalidatedAddress = UnvalidatedAddress of string

// 入力コマンド
type Command<'data> =
    { Data: 'data
      Timestamp: DateTime
      UserId: string }

type PlaceOrderCommand = Command<UnvalidatedOrder>

// パブリックAPI
type OrderPlaced = Undefined
type BillableOrderPlaced = Undefined
type AcknowledgmentSent = Undefined

type PlaceOrderEvent =
    | OrderPlaced of OrderPlaced
    | BillableOrderPlaced of BillableOrderPlaced
    | AcknowledgmentSent of AcknowledgmentSent

type PlaceOrderError = Undefined
type AsyncResult<'success, 'failure> = Async<Result<'success, 'failure>>

type PlaceOrderWorkflow =
    PlaceOrderCommand // 入力コマンド
        -> AsyncResult<PlaceOrderEvent list, PlaceOrderError>

File: Scala/chap07/src/main/scala/Domain.scala
import cats.data.EitherT

import java.time.LocalDateTime
import scala.concurrent.Future

object Domain:
  // 入力データ
  case class UnvalidatedCustomer(
                                  name: String,
                                  email: String,
                                )

  case class UnvalidatedAddress(value: String)

  case class UnvalidatedOrder(
                               orderId: String,
                               customerInfo: UnvalidatedCustomer,
                               shippingAddress: UnvalidatedAddress,
                             )

  // 入力コマンド
  case class Command[A](data: A, timestamp: LocalDateTime, userId: String)

  case class PlaceOrderCommand(value: Command[UnvalidatedOrder])

  // パブリックAPI
  enum PlaceOrderEvent:
    case OrderPlaced
    case BillableOrderPlaced
    case AcknowledgmentSent

  case class PlaceOrderError(value: Any)

  type PlaceOrderWorkflow =
    PlaceOrderCommand // In
      => EitherT[Future, PlaceOrderError, Seq[PlaceOrderEvent]] // Out
